### PR TITLE
Adds more user friendly error messages in ParseLockfile

### PR DESCRIPTION
**Pull request by emimuresan**
Closes: #16

## Description

Added enhanced and more informative error messages in `ParseLockfile.js`. The following error types are supported:

#### NO_OPTIONS:

Programatically generated the scenario below, this isn't something that would happen normally:

```
$ lockfile-lint -p __tests__/fixtures/yarn-only-http.lock --allowed-hosts yarn --validate-https
ABORTING lockfile lint process due to error exceptions

Did not recive options for lockfile path or type.

Error at new ParsingError (stack trace...)

error: command failed with exit code 1
```

#### NO_PARSER_FOR_TYPE:

```
$ lockfile-lint -p __tests__/fixtures/yarn-only-http.lock --allowed-hosts yarn --validate-https --type pnpm
ABORTING lockfile lint process due to error exceptions

Unable to find relevant lockfile parser for type "pnpm", the currently available options are npm,yarn.

Error at new ParsingError (stack trace...)

error: command failed with exit code 1
```

#### NO_PARSER_FOR_PATH:

```
$ lockfile-lint -p __tests__/fixtures/yarn-only-http.lock --allowed-hosts yarn --validate-https
ABORTING lockfile lint process due to error exceptions

Unable to find relevant lockfile parser for "__tests__/fixtures/yarn-only-http.lock", consider passing the --type option.

Error at new ParsingError (stack trace...)

error: command failed with exit code 1
```

#### READ_FAILED:

```
$ lockfile-lint -p __tests__/fixtures/non-existing-yarn.lock --allowed-hosts yarn --validate-https --type yarn
ABORTING lockfile lint process due to error exceptions

Unable to read lockfile "__tests__/fixtures/non-existing-yarn.lock"

Error: ENOENT: no such file or directory, open '__tests__/fixtures/non-existing-yarn.lock'
    at Object.openSync (fs.js:443:3)
    at Object.readFileSync (fs.js:343:35) (stack trace...)

error: command failed with exit code 1
```

#### PARSE_NPMLOCKFILE_FAILED:

```
$ lockfile-lint -p __tests__/fixtures/bad-package-lock.json --allowed-hosts yarn --validate-https --type npm
ABORTING lockfile lint process due to error exceptions

Unable to parse npm lockfile "__tests__/fixtures/bad-package-lock.json"

SyntaxError: Unexpected token m in JSON at position 23 (stack trace...)

error: command failed with exit code 1

```

#### PARSE_YARNLOCKFILE_FAILED

```
$ lockfile-lint -p __tests__/fixtures/bad-yarn.lock --allowed-hosts yarn --validate-https --type yarn
ABORTING lockfile lint process due to error exceptions

Unable to parse yarn lockfile "__tests__/fixtures/bad-yarn.lock"

SyntaxError: Unknown token: { line: 1, col: 0, type: 'INVALID', value: undefined } 1:0 in lockfile (stack trace...)

error: command failed with exit code 1

```

#### More Details

These error types are defined in `packages/lockfile-lint-api/src/common/ParsingError.js` and are used in `packages/lockfile-lint-api/src/ParseLockfile.js`.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Related Issue

#16 

## Motivation and Context

More descriptive/informative error messages when parsing a lockfile fails.

## How Has This Been Tested?

Added/Updated unit tests for all scenarios.

## Checklist:

- [x] I have updated the documentation (if required).
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I added a picture of a cute animal cause it's fun

giphy-cat-type

### CODE REVIEW

**Comment by emimuresan** [comment]
Not sure why the tests in TravisCI are failing with:

```
Expected substring: "expecting options object"
lockfile-lint-api:     Received message:   "Did not recive options for lockfile path or type"
```

I modified `parseLockfile.test`, so instead of "expecting options object" it actually should be "Did not recive options for lockfile path or type", and that is what it asserts on in the test file.

File: packages/lockfile-lint-api/src/common/ParsingError.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  NO_PARSER_FOR_PATH: path =>
+    `Unable to find relevant lockfile parser for "${path}", consider passing the --type option.`,
+  READ_FAILED: path => `Unable to read lockfile "${path}"`,
+  PARSE_NPMLOCKFILE_FAILED: path => `Unable to parse npm lockfile "${path}"`,
+  PARSE_YARNLOCKFILE_FAILED: path => `Unable to parse yarn lockfile "${path}"`
+}
+
+const ERROR_KEYS = Object.keys(errorMessages).reduce(function (accumulator, key) {
+  return Object.assign(accumulator, {[key]: key})
+}, {})
```
Comment (lirantal) [inline packages/lockfile-lint-api/src/common/ParsingError.js]: why not to just export `errorMessages` then?

File: packages/lockfile-lint-api/src/common/constants.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+'use strict'
+
+module.exports = {
```
Comment (lirantal) [inline packages/lockfile-lint-api/src/common/constants.js]: Great job here with moving it out to constants 👌

**Comment by lirantal** [comment]
@emimuresan I'm unable to reproduce the build failure locally either and I also tried resetting the build cache which didn't help. What about if you just open a new PR with the exact same code? I think you can just use a new branch to trigger a new PR

File: packages/lockfile-lint-api/__tests__/parseLockfile.test.js [hunk 3]
```diff
@@ -4,15 +4,46 @@
...
+    expect(() => {
+      new ParseLockfile({lockfilePath: '/path/to/file', lockfileType: 'npm'})
+    }).not.toThrowError()
+  })
+
   it('instantiating a parser with invalid object throws an error', () => {
     expect(() => {
       new ParseLockfile()
-    }).toThrowError('expecting options object')
+    }).toThrowError('Did not recive options for lockfile path or type')
```
Comment (lirantal) [inline packages/lockfile-lint-api/__tests__/parseLockfile.test.js]: typo, should be `receive`

File: packages/lockfile-lint-api/__tests__/parseLockfile.test.js [hunk 4]
```diff
@@ -4,15 +4,46 @@
...
   })
 
   it('instantiating a parser with string type throws an error', () => {
     expect(() => {
       new ParseLockfile('/path/to/file')
-    }).toThrowError('expecting options object')
+    }).toThrowError('Did not recive options for lockfile path or type')
```
Comment (lirantal) [inline packages/lockfile-lint-api/__tests__/parseLockfile.test.js]: typo, should be `receive`

File: packages/lockfile-lint-api/src/common/ParsingError.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
+'use strict'
+
+const {LOCKFILE_TYPES} = require('./constants')
+
+const errorMessages = {
+  NO_OPTIONS: () => 'Did not recive options for lockfile path or type',
```
Comment (lirantal) [inline packages/lockfile-lint-api/src/common/ParsingError.js]: typo, should be `receive`

**Comment by lirantal** [comment]
@emimuresan maybe you can rebase with master? checking out locally I see that I'm finding the test file being unstaged. This is probably why it's not getting through to the CI.

File: packages/lockfile-lint-api/src/common/ParsingError.js (hunk 1 above)
Comment (emimuresan) [reply packages/lockfile-lint-api/src/common/ParsingError.js]: i was overcomplicating things. now changed. thanks!

File: packages/lockfile-lint-api/__tests__/parseLockfile.test.js (hunk 3 above)
Comment (emimuresan) [reply packages/lockfile-lint-api/__tests__/parseLockfile.test.js]: fixed

File: packages/lockfile-lint-api/__tests__/parseLockfile.test.js (hunk 4 above)
Comment (emimuresan) [reply packages/lockfile-lint-api/__tests__/parseLockfile.test.js]: fixed

File: packages/lockfile-lint-api/src/common/ParsingError.js (hunk 5 above)
Comment (emimuresan) [reply packages/lockfile-lint-api/src/common/ParsingError.js]: fixed

**Comment by emimuresan** [comment]
> @emimuresan maybe you can rebase with master? checking out locally I see that I'm finding the test file being unstaged. This is probably why it's not getting through to the CI.

Yep, it was strange. Upon further investigation i discovered that there were 2 duplicate test files for ParseLockfile, i checked on master and it was the same:
```
parseLockFile.test.js
parseLockfile.test.js
```

I deleted `parseLockFile.test.js` in this PR. Hopefully it should be ok in CI now.

File: packages/lockfile-lint-api/__tests__/__fixtures__/bad-package-lock.json [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
+    "cat-generator": meow
```
[the remark this answers is not in the document]
Comment (emimuresan) [reply packages/lockfile-lint-api/__tests__/__fixtures__/bad-package-lock.json]: Yes, it's supposed to throw an error. I use it in one of the newly added unit tests for ParseLockfile.js

**Comment by lirantal** [comment]
Great work here @emimuresan ✨, thank you ❤️

**Pull request merged**